datafusion/optimizer/src/optimizer.rs

```diff
@@ -279,10 +294,10 @@ impl Optimizer {
     /// invoking observer function after each call
     pub fn optimize<F>(
         &self,
-        plan: &LogicalPlan,
+        plan: LogicalPlan,
```

Comment:
rewriting the optimizer passes to take in owned `LogicalPlan` is the key idea

I haven't plumbed it completely through yet, but the approach is looking promising